core/src/client/mod.rs

```diff
@@ -203,12 +203,25 @@ impl<Notif> Subscription<Notif> {
 		notifs_rx: mpsc::Receiver<JsonValue>,
 		kind: SubscriptionKind,
 	) -> Self {
-		Self { to_back, notifs_rx, kind, marker: PhantomData }
+		Self { to_back, notifs_rx, kind: Some(kind), marker: PhantomData }
 	}
 
 	/// Return the subscription type and, if applicable, ID.
 	pub fn kind(&self) -> &SubscriptionKind {
-		&self.kind
+		self.kind.as_ref().expect("only None after unsubscribe; qed")
+	}
+
+	/// Unsubscribe and consume the subscription.
+	pub async fn unsubscribe(mut self) -> Result<(), Error> {
```

Comment:
Is there a sensible way to test this?

Reply:
I think so, sorry but stuck on my other middleware PR :)